File: backend/src/main/resources/stubs/sms_service.py
"""
AEGIS DISPATCH - Enterprise SMS Gateway Integration (Python / FastAPI Reference Implementation)
Provides HIPAA-compliant pre-arrival alerts from Driver MDT to Hospital ED.

Dependencies:
    pip install fastapi uvicorn twilio pydantic
Run:
    uvicorn sms_service:app --port 8001 --reload
"""

import os
import datetime
from typing import Optional, Dict, Any
from fastapi import FastAPI, Form, HTTPException, Response
from pydantic import BaseModel
from twilio.rest import Client
from twilio.twiml.messaging_response import MessagingResponse

app = FastAPI(title="AEGIS CAD - SMS Notification Gateway")

TWILIO_ACCOUNT_SID = os.getenv("TWILIO_ACCOUNT_SID", "AC_DEMO_SID")
TWILIO_AUTH_TOKEN = os.getenv("TWILIO_AUTH_TOKEN", "demo_auth_token")
TWILIO_PHONE_NUMBER = os.getenv("TWILIO_PHONE_NUMBER", "+15005550006")

twilio_client = None
if TWILIO_ACCOUNT_SID and not TWILIO_ACCOUNT_SID.startswith("AC_DEMO"):
    twilio_client = Client(TWILIO_ACCOUNT_SID, TWILIO_AUTH_TOKEN)


class NotifyHospitalRequest(BaseModel):
    missionId: str
    hospitalId: str
    hospitalPhone: Optional[str] = "+918022220001"
    registrationNumber: Optional[str] = "KA-01-AE-1001"
    unitType: Optional[str] = "ALS"
    etaMinutes: Optional[float] = 8.0
    priority: Optional[str] = "CRITICAL"
    chiefComplaint: Optional[str] = "TRAUMA"
    vitals: Optional[Dict[str, Any]] = None


def build_hipaa_payload(req: NotifyHospitalRequest) -> str:
    """
    HIPAA / DPDP Compliance Enforcement:
    Strictly forbids full patient names, personal contact numbers, and explicit addresses.
    """
    patient_tag = f"PT-{req.missionId[:4].upper()}" if req.missionId else "PT-DEMO"
    
    vitals_text = "HR 84 | BP 124/82 | SpO2 98%"
    if req.vitals:
        hr = req.vitals.get("heartRate", 80)
        sys = req.vitals.get("bpSys", 120)
        dia = req.vitals.get("bpDia", 80)
        spo2 = req.vitals.get("spO2", 98)
        vitals_text = f"HR {hr} | BP {sys}/{dia} | SpO2 {spo2}%"

    return (
        f"[AEGIS-ALERT] INBOUND AMBULANCE: {req.registrationNumber} ({req.unitType})\n"
        f"ETA: ~{round(req.etaMinutes)} mins\n"
        f"PATIENT ID: #{patient_tag}\n"
        f"ACUITY: {req.priority}\n"
        f"CHIEF COMPLAINT: {req.chiefComplaint}\n"
        f"VITALS: {vitals_text}\n"
        f"REPLY:\n"
        f"  1 to CONFIRM TRAUMA BAY READY\n"
        f"  2 for ED DIVERSION (REROUTE)"
    )


@app.post("/api/missions/notify-hospital")
async def notify_hospital(payload: NotifyHospitalRequest):
    """
    Outbound SMS Dispatcher:
    Sends formatted clinical alert to designated emergency receiving department.
    """
    sms_text = build_hipaa_payload(payload)
    target_phone = payload.hospitalPhone or "+918022220001"

    message_sid = f"SIMULATED-{datetime.datetime.now().timestamp()}"
    if twilio_client:
        try:
            msg = twilio_client.messages.create(
                body=sms_text,
                from_=TWILIO_PHONE_NUMBER,
                to=target_phone
            )
            message_sid = msg.sid
        except Exception as e:
            print(f"[TWILIO ERROR] {e}. Falling back to simulated carrier.")

    print(f"[SMS LOG] Dispatched to {target_phone} (SID: {message_sid}):\n{sms_text}")

    return {
        "success": True,
        "messageSid": message_sid,
        "recipient": target_phone,
        "sentAt": datetime.datetime.utcnow().isoformat() + "Z",
        "payloadPreview": sms_text
    }


@app.post("/api/missions/sms-webhook")
async def sms_webhook(From: str = Form(...), Body: str = Form(...)):
    """
    2-Way Inbound Reply Webhook:
    Enables ED charge nurses to reply "1" to confirm trauma bay readiness or "2" to trigger auto-diversion.
    """
    reply = Body.strip()
    resp = MessagingResponse()

    if reply == "1":
        resp.message("[AEGIS CONFIRMATION] Trauma Bay 1 locked for inbound unit. Triage team alerted.")
        print(f"[WEBHOOK] Facility {From} confirmed Trauma Bay ready.")
    elif reply == "2":
        resp.message("[AEGIS DIVERSION] Diversion acknowledged. Central CAD auto-rerouting inbound ambulance.")
        print(f"[WEBHOOK] Facility {From} declared emergency diversion.")
    else:
        resp.message("AEGIS CAD: Reply 1 to Confirm Trauma Bay Ready, or 2 to declare Diversion.")

    return Response(content=str(resp), media_type="application/xml")

File: backend/src/main/resources/stubs/smsService.js
/**
 * AEGIS DISPATCH - Enterprise SMS Gateway Integration (Node.js Reference Implementation)
 * Provides HIPAA-compliant pre-arrival alerts from Driver MDT to Hospital ED.
 * 
 * Dependencies:
 *   npm install express twilio cors
 */

const express = require('express');
const twilio = require('twilio');

const app = express();
app.use(express.json());
app.use(express.urlencoded({ extended: false }));

const TWILIO_ACCOUNT_SID = process.env.TWILIO_ACCOUNT_SID || 'AC_DEMO_SID';
const TWILIO_AUTH_TOKEN = process.env.TWILIO_AUTH_TOKEN || 'demo_auth_token';
const TWILIO_PHONE_NUMBER = process.env.TWILIO_PHONE_NUMBER || '+15005550006';

const client = (TWILIO_ACCOUNT_SID && TWILIO_AUTH_TOKEN)
  ? twilio(TWILIO_ACCOUNT_SID, TWILIO_AUTH_TOKEN)
  : null;

/**
 * Build HIPAA / DPDP Compliant SMS Payload
 * STRICT REQUIREMENT: No patient names or street addresses over standard SMS.
 */
function buildHipaaPayload({ registrationNumber, unitType, etaMinutes, priority, chiefComplaint, patientId, vitals }) {
  const patientTag = patientId ? `PT-${patientId.substring(0, 4).toUpperCase()}` : 'PT-DEMO';
  const vitalsText = vitals
    ? `HR ${vitals.heartRate || 80} | BP ${vitals.bpSys || 120}/${vitals.bpDia || 80} | SpO2 ${vitals.spO2 || 98}%`
    : 'HR 84 | BP 124/82 | SpO2 98%';

  return (
    `[AEGIS-ALERT] INBOUND AMBULANCE: ${registrationNumber} (${unitType || 'ALS'})\n` +
    `ETA: ~${Math.round(etaMinutes)} mins\n` +
    `PATIENT ID: #${patientTag}\n` +
    `ACUITY: ${priority || 'CRITICAL'}\n` +
    `CHIEF COMPLAINT: ${chiefComplaint || 'CARDIAC'}\n` +
    `VITALS: ${vitalsText}\n` +
    `REPLY:\n` +
    `  1 to CONFIRM TRAUMA BAY READY\n` +
    `  2 for ED DIVERSION (REROUTE)`
  );
}

/**
 * Outbound Route: POST /api/missions/notify-hospital
 */
app.post('/api/missions/notify-hospital', async (req, res) => {
  try {
    const { missionId, hospitalId, hospitalPhone, registrationNumber, unitType, etaMinutes, priority, chiefComplaint, vitals } = req.body;

    const payload = buildHipaaPayload({
      registrationNumber: registrationNumber || 'KA-01-AE-1001',
      unitType: unitType || 'ALS',
      etaMinutes: etaMinutes || 8,
      priority: priority || 'CRITICAL',
      chiefComplaint: chiefComplaint || 'TRAUMA',
      patientId: missionId,
      vitals,
    });

    const targetPhone = hospitalPhone || '+918022220001';

    let messageSid = 'SIMULATED-' + Date.now();
    if (client && !TWILIO_ACCOUNT_SID.includes('DEMO')) {
      const message = await client.messages.create({
        body: payload,
        from: TWILIO_PHONE_NUMBER,
        to: targetPhone,
      });
      messageSid = message.sid;
    }

    console.log(`[SMS GATEWAY] Pre-Arrival SMS dispatched to ${targetPhone} (SID: ${messageSid})`);
    return res.status(200).json({
      success: true,
      messageSid,
      sentAt: new Date().toISOString(),
      recipient: targetPhone,
      payloadPreview: payload,
    });
  } catch (err) {
    console.error('[SMS ERROR]', err);
    return res.status(500).json({ error: err.message });
  }
});

/**
 * Inbound 2-Way Webhook: POST /api/missions/sms-webhook
 * Triggered when hospital charge nurse replies "1" (Accept) or "2" (Divert)
 */
app.post('/api/missions/sms-webhook', (req, res) => {
  const replyBody = (req.body.Body || '').trim();
  const fromNumber = req.body.From;

  const twiml = new twilio.twiml.MessagingResponse();

  if (replyBody === '1') {
    twiml.message('[AEGIS CONFIRMATION] Trauma Bay 1 locked for inbound unit. Triage team alerted.');
    console.log(`[WEBHOOK] Facility ${fromNumber} confirmed Trauma Bay ready.`);
  } else if (replyBody === '2') {
    twiml.message('[AEGIS DIVERSION] Diversion acknowledged. Central CAD auto-rerouting inbound ambulance.');
    console.log(`[WEBHOOK] Facility ${fromNumber} declared emergency diversion.`);
  } else {
    twiml.message('AEGIS CAD Command: Reply 1 to Confirm Trauma Bay Ready, or 2 to declare Diversion.');
  }

  res.writeHead(200, { 'Content-Type': 'text/xml' });
  res.end(twiml.toString());
});

const PORT = process.env.PORT || 3001;
app.listen(PORT, () => {
  console.log(`AEGIS Node.js SMS Service running on port ${PORT}`);
});
